### PR TITLE
IGNITE-12148: [ML] Recommendation Engine

### PR DESCRIPTION
**Pull request by zaleslaw**

### CODE REVIEW

**Comment by zaleslaw** [comment]
Add updatable Recommendation Engine + support for ModelStorage in Windows

**Comment by zaleslaw** [comment]
@avplatonov Please make review

File: modules/ml/pom.xml [hunk 1]
```diff
@@ -102,7 +102,6 @@
             <groupId>org.springframework</groupId>
             <artifactId>spring-context</artifactId>
             <version>${spring.version}</version>
-            <scope>test</scope>
```
Comment (YuriBabak) [inline modules/ml/pom.xml]: We really need this change?

File: modules/ml/src/main/java/org/apache/ignite/ml/recommendation/RecommendationModel.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import org.apache.ignite.ml.IgniteModel;
+import org.apache.ignite.ml.math.primitives.vector.Vector;
+
+/**
+ * Recommendation model that predicts rating for {@link ObjectSubjectPair}.
+ *
+ * @param <O> Type of an object of recommendation.
+ * @param <S> Type of a subject of recommendation.
+ */
+// TODO: GG-22915 Investigate ability to distribute recommendation model
```
Comment (YuriBabak) [inline modules/ml/src/main/java/org/apache/ignite/ml/recommendation/RecommendationModel.java]: Please, cleanup this

File: modules/ml/pom.xml (hunk 1 above)
Comment (zaleslaw) [reply modules/ml/pom.xml]: yep, we use spring in the resolving resource path

File: modules/ml/src/main/java/org/apache/ignite/ml/recommendation/RecommendationModel.java (hunk 2 above)
Comment (zaleslaw) [reply modules/ml/src/main/java/org/apache/ignite/ml/recommendation/RecommendationModel.java]: Agree

File: examples/src/main/java/org/apache/ignite/examples/ml/recommendation/MovieLensExample.java [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                    .withLearningRate(10)
+                    .withLearningEnvironmentBuilder(envBuilder)
+                    .withTrainerEnvironment(envBuilder.buildForTrainer());
+
+                RecommendationModel<Integer, Integer> mdl = trainer.fit(
+                    new CacheBasedDatasetBuilder<>(ignite, movielensCache)
+                );
+
+                double mean = 0;
+                try (QueryCursor<Cache.Entry<Integer, RatingPoint>> cursor = movielensCache.query(new ScanQuery<>())) {
```
Comment (avplatonov) [inline examples/src/main/java/org/apache/ignite/examples/ml/recommendation/MovieLensExample.java]: We can compute RSS using just one query, there is no need to do two-pass process of estimation. Please, have a look at this page: https://en.wikipedia.org/wiki/Variance . Variance could be computed using two means - mean value of original data and mean value of squared data. We can compute these two values to determine variance that used in RSS formula.

**Pull request merged**